### PR TITLE
[FLINK-20072][docs] Add documentation for FLIP-107

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Documents all features of FLIP-107.

## Brief change log

Updates:
- DDL
- Kafka connector
- Upsert Kafka connector
- Debezium format

Fixes various issues with:
- Kinesis connector
- Format page
- Connector page

Additionally, fixes nullability in Debezium metadata.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/dev/table/connectors/index.md [hunk 1]
```diff
@@ -95,20 +95,23 @@ Flink natively support various connectors. The following tables list all availab
 How to use connectors
 --------
 
-Flink supports to use SQL CREATE TABLE statement to register a table. One can define the table name, the table schema, and the table options for connecting to an external system.
+Flink supports to use SQL `CREATE TABLE` statement to register a table. One can define the table name,
+the table schema, and the table options for connecting to an external system.
```
Comment (sjwiesman) [inline docs/dev/table/connectors/index.md:99]:
```suggestion
Flink supports using SQL `CREATE TABLE` statements to register tables. One can define the table name,
the table schema, and the table options for connecting to an external system.
```

File: docs/dev/table/sql/create.md [hunk 2]
```diff
@@ -182,24 +187,165 @@ CREATE TABLE [catalog_name.][db_name.]table_name
...
 
-**COMPUTED COLUMN**
+**Physical / Regular Columns**
 
-A computed column is a virtual column that is generated using the syntax  "`column_name AS computed_column_expression`". It is generated from a non-query expression that uses other columns in the same table and is not physically stored within the table. For example, a computed column could be defined as `cost AS price * quantity`. The expression may contain any combination of physical column, constant, function, or variable. The expression cannot contain a subquery.
+Physical columns are regular columns known from databases. They define the names, the types, and the
+order of fields in the physical data. Thus, physical columns represent the payload that is read from
+and written to an external system. Connectors and formats are using these columns (in the defined order)
+to configure themselves. Other kinds of columns can be declared between physical columns but will not
+influence the final physical schema.
```
Comment (sjwiesman) [inline docs/dev/table/sql/create.md:201]:
```suggestion
Physical columns are regular columns known from databases. They define the names, the types, and the
order of fields in the physical data. Thus, physical columns represent the payload that is read from
and written to an external system. Connectors and formats use these columns (in the defined order)
to configure themselves. Other kinds of columns can be declared between physical columns but will not
influence the final physical schema.
```

File: docs/dev/table/sql/create.md [hunk 3]
```diff
@@ -182,24 +187,165 @@ CREATE TABLE [catalog_name.][db_name.]table_name
...
+  `name` STRING,
+  `timestamp` TIMESTAMP(3) WITH LOCAL TIME ZONE METADATA    -- use column name as metadata key
+) WITH (
+  'connector' = 'kafka'
+  ...
+);
+{% endhighlight %}
+
+For convenience, the runtime will perform an explicit cast if the data type of the column differs from
+the data type of the metadata field. Of course this requires that the two data types are compatible.
```
Comment (sjwiesman) [inline docs/dev/table/sql/create.md:261]:
```suggestion
the data type of the metadata field. Of course, this requires that the two data types are compatible.
```

File: docs/dev/table/sql/create.md [hunk 4]
```diff
@@ -182,24 +187,165 @@ CREATE TABLE [catalog_name.][db_name.]table_name
...
+MyTable(`timestamp` BIGINT, `user_id` BIGINT, `name` STRING)
+{% endhighlight %}
+
+**Computed Columns**
+
+Computed columns are virtual columns that are generated using the syntax `column_name AS computed_column_expression`.
+
+A computed column evaluates an expression that can reference other columns declared in the same table.
+Both physical columns and metadata columns can be accessed if they preceed the computed column in the
+schema declaration. The column itself is not physically stored within the table. The column's data type
```
Comment (sjwiesman) [inline docs/dev/table/sql/create.md:307]: Can you clarify when the computed column is evaluated? At the source or on demand? Or both?

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/debezium/DebeziumJsonDecodingFormat.java [hunk 5]
```diff
@@ -150,13 +150,16 @@ public Object convert(GenericRowData row, int pos) {
 
 		INGESTION_TIMESTAMP(
 			"ingestion-timestamp",
-			DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(3).notNull(),
+			DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(3).nullable(),
```
Comment (sjwiesman) [inline flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/debezium/DebeziumJsonDecodingFormat.java:153]: Is this supposed to be part of this PR?

File: docs/dev/table/connectors/kafka.md [hunk 6]
```diff
@@ -189,6 +362,130 @@ Connector Options
...
+
+  'value.format' = 'json',
+  'value.json.fail-on-missing-field' = 'false',
+  'value.fields-include' = 'ALL'
+)
+{% endhighlight %}
+</div>
+</div>
+
+The key format includes the fields listed in `'key.fields'` (in the same order). Thus, it will be configured with
```
Comment (sjwiesman) [inline docs/dev/table/connectors/kafka.md:431]: Can you specify the delimiter? I'm guessing `;` based on the example but I'm not positive.

File: docs/dev/table/connectors/kinesis.md [hunk 7]
```diff
@@ -85,44 +82,46 @@ The following metadata can be exposed as read-only (`VIRTUAL`) columns in a tabl
...
- 'stream' = 'user_behavior',
- 'aws.region' = 'us-east-2',
- 'scan.stream.initpos' = 'LATEST',
- 'format' = 'csv'
+  `user_id` BIGINT,
+  `item_id` BIGINT,
+  `category_id` BIGINT,
+  `behavior` STRING,
+  `ts` TIMESTAMP(3),
+  `arrival_time` TIMESTAMP(3) METADATA FROM 'timestamp' VIRTUAL,
```
Comment (sjwiesman) [inline docs/dev/table/connectors/kinesis.md:113]: On the debezium page you said only Kafka can forward metadata. Is that not true, or did I misunderstand what you said.

File: docs/dev/table/connectors/upsert-kafka.md [hunk 8]
```diff
@@ -181,6 +199,12 @@ Connector Options
 Features
 ----------------
 
+### Key and Value Formats
+
+See the [regular Kafka connector]({% link dev/connectors/kafka.md %}#key-and-value-formats) for more
+explanation around key and value formats. However, note that this connector requires both a key and
+value format where the key fields are derived from the `PRIMARY KEY` constraint.
```
Comment (sjwiesman) [inline docs/dev/table/connectors/upsert-kafka.md:206]: Can you add an example, I suspect this will be a very common use case.

**Review by sjwiesman — CHANGES_REQUESTED** [review]

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/debezium/DebeziumJsonDecodingFormat.java (hunk 5 above)
Comment (twalthr) [reply flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/debezium/DebeziumJsonDecodingFormat.java:153]: I spotted this while going through the docs of Debezium and corrected it inline.

File: docs/dev/table/connectors/kafka.md (hunk 6 above)
Comment (twalthr) [reply docs/dev/table/connectors/kafka.md:431]: The delimiter is a regular Flink string lists delimiter. But I can also mention it explicitly.

File: docs/dev/table/connectors/kinesis.md (hunk 7 above)
Comment (twalthr) [reply docs/dev/table/connectors/kinesis.md:113]: It's about format metadata. Currently, the Kinesis connector does not forward metadata coming from Debezium. I will open an issue for this btw.

Comment (twalthr) [reply docs/dev/table/connectors/kinesis.md:113]: I also improved the docs a bit.

**Comment by twalthr** [comment]
Thanks @sjwiesman. I hope I could address all you comments.

**Review by sjwiesman — APPROVED** [review]
LGTM, feel free to merge

**Pull request closed without merging**